File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeNotebook.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Wisp.ProjectBrowser.Contracts;

public sealed record NativeNotebookCell(int Index, string Language, string Source, string Output, bool? Ok, string Origin)
{
    public string StarKey => Language + "\0" + Source;
    public string Status => Ok == true ? "ok" : Ok == false ? "error" : Origin == "assistant" ? "source" : "running";
    public bool OutputInitiallyExpanded => Ok == false;
    public static NativeNotebookCell[] Collect(IEnumerable<ConversationItem> items)
    {
        var cells = new List<NativeNotebookCell>();
        foreach (var item in items)
        {
            if (item.Role == "assistant")
            {
                foreach (var (language, source) in Fences(item.Text))
                    if (language is not ("csv" or "tsv" or "fasta" or "fa")) cells.Add(new(0, language.Length == 0 ? "text" : language, source, "", null, "assistant"));
            }
            else if (item.Role == "tool" && item.ToolName is "python" or "r" or "shell" && !string.IsNullOrWhiteSpace(item.Input))
            {
                var source = item.Input;
                var end = source.IndexOf("] ", StringComparison.Ordinal);
                if (item.ToolName != "shell" && source.StartsWith('[') && end >= 0) source = source[(end + 2)..];
                cells.Add(new(0, item.ToolName == "shell" ? "bash" : item.ToolName, source, item.Text, item.Ok, item.ToolName == "shell" ? "shell" : "repl"));
            }
        }
        var executed = cells.Where(cell => cell.Origin != "assistant").Select(cell => cell.Source.Trim()).ToHashSet();
        return cells.Where(cell => cell.Origin != "assistant" || !executed.Contains(cell.Source.Trim())).Select((cell, index) => cell with { Index = index }).ToArray();
    }
    private static IEnumerable<(string Language, string Source)> Fences(string text)
    {
        var lines = text.Split('\n').Select(line => line.EndsWith('\r') ? line[..^1] : line).ToList();
        if (text.EndsWith('\n')) lines.RemoveAt(lines.Count - 1);
        for (var index = 0; index < lines.Count; index++)
        {
            var fence = lines[index].Trim();
            if (!fence.StartsWith("```", StringComparison.Ordinal)) continue;
            var raw = fence.TrimStart('`').Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault() ?? "";
            var language = new string(raw.Select(c => c is >= 'A' and <= 'Z' ? (char)(c + 32) : c).ToArray());
            var end = index + 1;
            while (end < lines.Count && !lines[end].Trim().StartsWith("```", StringComparison.Ordinal)) end++;
            var source = string.Join("\n", lines.Skip(index + 1).Take(end - index - 1));
            if (source.Length > 0) yield return (language, source);
            index = end;
        }
    }
}
public sealed record NativeNotebookStar(string Id, string Kind, string? Language, string Code, string SourceProjectId, string SourceSessionId)
{
    public bool Belongs(string project, string session) => Kind == "code" && SourceProjectId == project && SourceSessionId == session;
    public bool Matches(NativeNotebookCell cell) => (Language ?? "") == cell.Language && Code == cell.Source;
}
public interface INativeNotebookClient
{
    Task<NativeNotebookStar[]> ListStarsAsync(string project, string session, CancellationToken token = default);
    Task<NativeNotebookStar> StarAsync(string project, string session, string language, string code, CancellationToken token = default);
    Task<bool> UnstarAsync(string project, string session, string id, CancellationToken token = default);
}
public sealed class NativeNotebookClient(INativeSettingsClient transport) : INativeNotebookClient
{
    public async Task<NativeNotebookStar[]> ListStarsAsync(string project, string session, CancellationToken token = default)
    {
        var value = await transport.InvokeAsync("native_conversation_panel_notebook_stars", new JsonObject { ["session_id"] = session }, project, token).ConfigureAwait(false);
        var rows = value?.Deserialize<NativeNotebookStar[]>(ConversationSnapshot.JsonOptions) ?? throw new InvalidDataException("Missing notebook stars");
        if (rows.Any(row => !row.Belongs(project, session))) throw new InvalidDataException("Notebook star scope mismatch");
        return rows;
    }
    public async Task<NativeNotebookStar> StarAsync(string project, string session, string language, string code, CancellationToken token = default)
    {
        var value = await transport.InvokeAsync("native_conversation_panel_notebook_star", new JsonObject { ["session_id"] = session, ["language"] = language, ["code"] = code }, project, token).ConfigureAwait(false);
        var row = value?.Deserialize<NativeNotebookStar>(ConversationSnapshot.JsonOptions) ?? throw new InvalidDataException("Missing saved code");
        if (!row.Belongs(project, session) || row.Language != language || row.Code != code) throw new InvalidDataException("Saved code identity mismatch");
        return row;
    }
    public async Task<bool> UnstarAsync(string project, string session, string id, CancellationToken token = default) =>
        (await transport.InvokeAsync("native_conversation_panel_notebook_unstar", new JsonObject { ["session_id"] = session, ["library_item_id"] = id }, project, token).ConfigureAwait(false))?.GetValue<bool>() ?? throw new InvalidDataException("Missing removal result");
}

File: apps/macos/Sources/WispProjectBrowser/NativeNotebook.swift
import Foundation

/// A projection of displayed transcript code; never a second execution engine.
public struct NativeNotebookCell: Codable, Equatable, Identifiable, Sendable {
    public let index: Int
    public let language: String
    public let source: String
    public let output: String
    public let ok: Bool?
    public let origin: String
    public var id: Int { index }
    public var starKey: String { language + "\u{0}" + source }
    public var status: String { ok == true ? "ok" : ok == false ? "error" : origin == "assistant" ? "source" : "running" }
    public var outputInitiallyExpanded: Bool { ok == false }
    public static func collect(_ items: [ConversationItem]) -> [Self] {
        var cells: [Self] = []
        for item in items {
            if item.role == "assistant" {
                for (language, source) in fences(item.text) where !["csv", "tsv", "fasta", "fa"].contains(language) {
                    cells.append(Self(index: 0, language: language.isEmpty ? "text" : language, source: source, output: "", ok: nil, origin: "assistant"))
                }
            } else if item.role == "tool", let name = item.tool_name, ["python", "r", "shell"].contains(name), let input = item.input, !input.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                var source = input
                if name != "shell", input.hasPrefix("["), let range = input.range(of: "] ") { source = String(input[range.upperBound...]) }
                cells.append(Self(index: 0, language: name == "shell" ? "bash" : name, source: source, output: item.text, ok: item.ok, origin: name == "shell" ? "shell" : "repl"))
            }
        }
        let executed = Set(cells.filter { $0.origin != "assistant" }.map { $0.source.trimmingCharacters(in: .whitespacesAndNewlines) })
        return cells.filter { $0.origin != "assistant" || !executed.contains($0.source.trimmingCharacters(in: .whitespacesAndNewlines)) }.enumerated().map { index, cell in
            Self(index: index, language: cell.language, source: cell.source, output: cell.output, ok: cell.ok, origin: cell.origin)
        }
    }
    private static func fences(_ text: String) -> [(String, String)] {
        var lines = text.components(separatedBy: "\n").map { $0.hasSuffix("\r") ? String($0.dropLast()) : $0 }
        if text.hasSuffix("\n") { lines.removeLast() }
        var result: [(String, String)] = []; var index = 0
        while index < lines.count {
            let fence = lines[index].trimmingCharacters(in: .whitespacesAndNewlines)
            guard fence.hasPrefix("```") else { index += 1; continue }
            let raw = fence.drop(while: { $0 == "`" }).split(whereSeparator: { $0.isWhitespace }).first.map(String.init) ?? ""
            let language = String(String.UnicodeScalarView(raw.unicodeScalars.map { (65...90).contains($0.value) ? UnicodeScalar($0.value + 32)! : $0 }))
            var end = index + 1
            while end < lines.count && !lines[end].trimmingCharacters(in: .whitespacesAndNewlines).hasPrefix("```") { end += 1 }
            let source = lines[(index + 1)..<end].joined(separator: "\n")
            if !source.isEmpty { result.append((language, source)) }
            index = end + 1
        }
        return result
    }
}
public struct NativeNotebookStar: Codable, Equatable, Identifiable, Sendable {
    public let id: String
    public let kind: String
    public let language: String?
    public let code: String
    public let source_project_id: String
    public let source_session_id: String
    public func belongs(project: String, session: String) -> Bool { kind == "code" && source_project_id == project && source_session_id == session }
    public func matches(_ cell: NativeNotebookCell) -> Bool { (language ?? "") == cell.language && code == cell.source }
}
